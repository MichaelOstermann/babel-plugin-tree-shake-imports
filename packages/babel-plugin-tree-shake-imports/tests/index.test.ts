import type { Options } from "../src"
import { describe, expect, it } from "vitest"
import { expectSnapshot } from "./helpers"

const defaultOptions: Options = {
    resolve({ importPath, propertyName }) {
        return {
            path: `${importPath}/${propertyName}`,
            type: "named",
        }
    },
}

describe("babel-plugin-tree-shake-imports", () => {
    it("Should tree-shake named imports", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
        `, {
            resolve(data) {
                expect(data).toEqual({
                    filePath: "",
                    importName: "Foo",
                    importPath: "foo",
                    importType: "named",
                    localName: "Foo",
                    propertyName: "bar",
                })
                return defaultOptions.resolve(data)
            },
        })
    })

    it("Should tree-shake aliased imports", () => {
        expectSnapshot(`
            import { Foo as Bar } from "foo";
            Bar.bar;
        `, {
            resolve(data) {
                expect(data).toEqual({
                    filePath: "",
                    importName: "Foo",
                    importPath: "foo",
                    importType: "named",
                    localName: "Bar",
                    propertyName: "bar",
                })
                return defaultOptions.resolve(data)
            },
        })
    })

    it("Should tree-shake string-literal imports", () => {
        expectSnapshot(`
            import { "foo" as Foo } from "foo";
            Foo.bar;
        `, {
            resolve(data) {
                expect(data).toEqual({
                    filePath: "",
                    importName: "foo",
                    importPath: "foo",
                    importType: "named",
                    localName: "Foo",
                    propertyName: "bar",
                })
                return defaultOptions.resolve(data)
            },
        })
    })

    it("Should tree-shake default imports", () => {
        expectSnapshot(`
            import Foo from "foo";
            Foo.bar;
        `, {
            resolve(data) {
                expect(data).toEqual({
                    filePath: "",
                    importName: "Foo",
                    importPath: "foo",
                    importType: "default",
                    localName: "Foo",
                    propertyName: "bar",
                })
                return defaultOptions.resolve(data)
            },
        })
    })

    it("Should tree-shake wildcard imports", () => {
        expectSnapshot(`
            import * as Foo from "foo";
            Foo.bar;
        `, {
            resolve(data) {
                expect(data).toEqual({
                    filePath: "",
                    importName: "Foo",
                    importPath: "foo",
                    importType: "wildcard",
                    localName: "Foo",
                    propertyName: "bar",
                })
                return defaultOptions.resolve(data)
            },
        })
    })

    it("Should tree-shake multiple references", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
            Foo.bar;
            Foo.baz;
            Foo.baz;
        `, defaultOptions)
    })

    it("Should tree-shake into default import", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
        `, {
            resolve({ importPath, propertyName }) {
                return {
                    path: `${importPath}/${propertyName}`,
                    type: "default",
                }
            },
        })
    })

    it("Should tree-shake using alias", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar.baz;
        `, {
            resolve({ importPath, propertyName }) {
                return {
                    importName: "Bar",
                    path: `${importPath}/${propertyName}`,
                    type: "named",
                }
            },
        })
    })

    it("Should tree-shake nested imports", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar.baz;
        `, defaultOptions)
    })

    it("Should not tree-shake when not desired", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            import { Bar } from "bar";
            Foo.bar;
            Bar.baz;
        `, {
            resolve(data) {
                if (data.importName === "Bar") return
                return defaultOptions.resolve(data)
            },
        })
    })

    it("Should skip irrelevant references", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            function example() {
                const Foo = {};
                return Foo.bar;
            };
        `, defaultOptions)
    })

    it("Should skip fishy properties", () => {
        expectSnapshot(`
            import { Foo } from "foo";
            Foo.bar;
            Foo["bar"];
        `, defaultOptions)
    })

    it("Should keep unused imports", () => {
        expectSnapshot(`
            import { Foo, Bar } from "foo";
            Foo.bar;
        `, defaultOptions)
    })
})
