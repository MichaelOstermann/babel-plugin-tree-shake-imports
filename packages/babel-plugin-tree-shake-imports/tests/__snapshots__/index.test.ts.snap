// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`babel-plugin-tree-shake-imports > Should keep unused imports 1`] = `
"import { Bar } from "foo";
import { bar as _bar } from "foo/bar";
_bar;"
`;

exports[`babel-plugin-tree-shake-imports > Should not tree-shake when not desired 1`] = `
"import { bar as _bar } from "foo/bar";
import { Bar } from "bar";
_bar;
Bar.baz;"
`;

exports[`babel-plugin-tree-shake-imports > Should skip fishy properties 1`] = `
"import { Foo } from "foo";
import { bar as _bar } from "foo/bar";
_bar;
Foo["bar"];"
`;

exports[`babel-plugin-tree-shake-imports > Should skip irrelevant references 1`] = `
"import { Foo } from "foo";
function example() {
  const Foo = {};
  return Foo.bar;
}
;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake aliased imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake default imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake into default import 1`] = `
"import _bar from "foo/bar";
_bar;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake multiple references 1`] = `
"import { baz as _baz } from "foo/baz";
import { bar as _bar } from "foo/bar";
_bar;
_bar;
_baz;
_baz;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake named imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake nested imports 1`] = `
"import { baz as _baz } from "foo/bar/baz";
_baz;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake string-literal imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake using alias 1`] = `
"import { Bar as _baz } from "foo/bar/baz";
_baz;"
`;

exports[`babel-plugin-tree-shake-imports > Should tree-shake wildcard imports 1`] = `
"import { bar as _bar } from "foo/bar";
_bar;"
`;
